From: Benjamin Drung <benjamin.drung@canonical.com>
Date: Tue, 6 Jan 2026 00:32:20 +0100
Subject: fix(dracut): source dracut-functions.sh before calling dwarning

`dracut-functions.sh` sources `dracut-logger.sh` which defines the
`dwarning` function. `dracut.sh` might call `dwarning` before
`dracut-functions.sh` is sourced.

`dracut-functions.sh` calls `find_binary` and `dlog_init` when sourced.
`find_binary` needs `dracutsysrootdir` set and `dlog_init` needs
`stdloglvl`, `sysloglvl`, `kmsgloglvl`, `maxloglvl`, `fileloglvl`,
`logfile`. Move sourcing `dracut-functions.sh` as early as possible, but
after setting those variables.

Backport this change to release 109.

Follow-up for 8d9887b
Fixes: https://bugs.debian.org/1124479
Forwarded: https://github.com/dracut-ng/dracut-ng/pull/1993
---
 dracut.sh | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dracut.sh b/dracut.sh
index 7107ce6..3c76a97 100755
--- a/dracut.sh
+++ b/dracut.sh
@@ -1192,6 +1192,10 @@ drivers_dir="${drivers_dir%"${drivers_dir##*[!/]}"}"
 [[ $sbat_l ]] && sbat="$sbat_l"
 [[ $machine_id_l ]] && machine_id="$machine_id_l"
 
+# shellcheck source=./dracut-logger.sh
+. "$dracutbasedir/dracut-logger.sh"
+dlog_init
+
 if ! [[ $outfile ]]; then
     if [[ $machine_id != "no" ]]; then
         if [[ -d "${dracutsysrootdir-}"/efi/Default ]] \
